src/aggregation/bucket/range.rs

```diff
@@ -0,0 +1,437 @@
+use std::cmp::Ordering;
+use std::ops::Range;
+
+use itertools::Itertools;
+
+use crate::aggregation::agg_req_with_accessor::{
+    AggregationsWithAccessor, BucketAggregationWithAccessor,
+};
+use crate::aggregation::intermediate_agg_result::IntermediateBucketResult;
+use crate::aggregation::segment_agg_result::{
+    SegmentAggregationResultsCollector, SegmentBucketDataEntry, SegmentBucketDataEntryKeyCount,
+};
+use crate::aggregation::{f64_from_fastfield_u64, f64_to_fastfield_u64, Key};
+use crate::fastfield::FastFieldReader;
+use crate::schema::Type;
+use crate::{DocId, TantivyError};
+
+#[derive(Clone, Debug, PartialEq)]
+/// Provide user-defined buckets to aggregate on.
+/// Two special buckets will automatically be created to cover the whole range of values.
+/// The provided buckets have to be continous.
+/// During the aggregation, the values extracted from the fast_field `field_name` will be checked
+/// against each bucket range. Note that this aggregation includes the from value and excludes the
+/// to value for each range.
+///
+/// Result type is
+/// [BucketDataEntryKeyCount](crate::aggregation::agg_result::BucketDataEntryKeyCount)
+pub struct RangeAggregation {
+    /// The field to aggregate on.
+    pub field_name: String,
+    /// Note that this aggregation includes the from value and excludes the to value for each
+    /// range. Extra buckets will be created until the first to, and last from.
+    pub buckets: Vec<Range<f64>>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct SegmentRangeBucketEntry {
+    range: Range<u64>,
+    bucket: SegmentBucketDataEntry,
+}
+
+/// The collector puts values from the fast field into the correct buckets and does a conversion to
+/// the correct datatype.
+#[derive(Clone, Debug, PartialEq)]
+pub struct SegmentRangeCollector {
+    /// The buckets containing the aggregation data.
+    buckets: Vec<SegmentRangeBucketEntry>,
+    field_type: Type,
```

Comment:
I suspect it might be more efficient and proofreadable to have a `Vec<u64>` directly in the `SegmentRangeCollector`.

RIght now the property that the SegmentReangeBucketEntry do not overlap and has no gap is not expressed by the type.